django/core/cache/backends/memcached.py

```diff
@@ -160,19 +170,68 @@ def touch(self, key, timeout=DEFAULT_TIMEOUT, version=None):
         key = self.make_key(key, version=version)
         return self._cache.touch(key, self.get_backend_timeout(timeout)) != 0
 
-    def get(self, key, default=None, version=None):
-        key = self.make_key(key, version=version)
-        val = self._cache.get(key)
+    def __get(self, memcache, key, default):
+
         # python-memcached doesn't support default values in get().
+        # this method will help us in getting the default value if
+        # key doesn't exist, if the Key exists with None as value then
+        # it will return None instead of giving default value.
+        # It will help us in get_or_set to determine if we already
+        # have a key with None or not.
         # https://github.com/linsomniac/python-memcached/issues/159
         # Remove this method if that issue is fixed.
-        if val is None:
-            return default
+
+        key = memcache._encode_key(key)
+        if memcache.do_check_key:
+            memcache.check_key(key)
+        server, key = memcache._get_server(key)
+        if not server:
+            return None
+
+        def _unsafe_get():
+
+            try:
+                fullcmd = b''.join((b'get ', key))
+                server.send_cmd(fullcmd)
+                rkey, flags, rlen, = memcache._expectvalue(
+                    server, raise_exception=True
+                )
+
+                if not rkey:
+                    return default
+                try:
+                    value = memcache._recv_value(server, flags, rlen)
+                finally:
+                    server.expect(b"END", raise_exception=True)
+            except (_Error, socket.error) as msg:
+                if isinstance(msg, tuple):
+                    msg = msg[1]
+                server.mark_dead(msg)
+                return None
+
+            return value
+
+        try:
+            return _unsafe_get()
+        except _ConnectionDeadError:
+            # retry once
+            try:
+                if server.connect():
+                    return _unsafe_get()
+                return None
+            except (_ConnectionDeadError, socket.error) as msg:
+                server.mark_dead(msg)
+            return None
```

Comment:
I really don't understand why we're introducing a complicated low-level workaround. It looks over engineered. IMO it's not worth to support `None` values on `memcached` if there is no better way to handle them.

@pope1ni Can I ask for your opinion about this patch?

Reply:
@felixxm Will have a look at this tomorrow.

Reply:
@felixxm So the [proposed change](https://github.com/linsomniac/python-memcached/pull/158/files) to support a default argument for get in python-memcached is actually quite simple. But supporting it here in Django requires copying the whole thing, which is not very nice.

Unfortunately the last commit for python-memcached was nearly three years ago now. I've pinged @linsomniac again regarding the possibility of transferring maintainership to Jazzband ([comment](https://github.com/linsomniac/python-memcached/issues/95#issuecomment-725562601)), but if that doesn't proceed I can only suggest that we deprecate support for python-memcached in 3.2 now that we have support for pymemcache as an alternative. If it does move to Jazzband we can fix this issue.

As far as the handling of `None` as a value goes, there are a number of other cache operations that are treating `None` returned from the cache as the key not existing which should also be fixed. I've knocked together a quick thing as a PoC which I can tidy up if you think it looks like a sensible way forward: #13671. We basically just keep the existing behaviour for the python-memcached backend, but fix for all other backends.

Reply:
> @felixxm So the [proposed change](https://github.com/linsomniac/python-memcached/pull/158/files) to support a default argument for get in python-memcached is actually quite simple. But supporting it here in Django requires copying the whole thing, which is not very nice.

Yes, it's not.

> Unfortunately the last commit for python-memcached was nearly three years ago now. I've pinged @linsomniac again regarding the possibility of transferring maintainership to Jazzband ([comment](https://github.com/linsomniac/python-memcached/issues/95#issuecomment-725562601)), but if that doesn't proceed I can only suggest that we deprecate support for python-memcached in 3.2 now that we have support for pymemcache as an alternative. If it does move to Jazzband we can fix this issue.

I think it's time to deprecate it, I've created a separate ticket, see ticket-32193.

> As far as the handling of `None` as a value goes, there are a number of other cache operations that are treating `None` returned from the cache as the key not existing which should also be fixed. I've knocked together a quick thing as a PoC which I can tidy up if you think it looks like a sensible way forward: #13671. We basically just keep the existing behaviour for the python-memcached backend, but fix for all other backends.

Agreed, maybe with a small warning in docs.